### DISCUSSION TITLE
Cancelling rest of job if condition is met.

### DISCUSSION THREAD

**Discussion by BlackSwan2019** [community/community#25280] [Actions]
<p>When a developer opens a PR, we want to run a job only if code within certain files have changed. The issue is that in order to know whether to run the job or not, the job needs to be run first in order to checkout the repo to perform a git diff on it (kind of a chicken and egg)! So, having established that, I thought maybe we can let the job run. There would be a step (after the checkout@v1) that would then determine whether to continue the job execution or not. If not, then cancel rest of job. Is there a “cancel rest of job” command? We’d want it to also not be considered a failed job. Thanks! </p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/blackswan2019">@blackswan2019</a> ,</p>
<p>We have no the “cancel rest of job” command.<br>
As a workaround, you can try using the <strong>needs</strong> and <strong>if conditional</strong> on the subsequent jobs to meet your requests.<br>
A simple demo:</p>
<pre><code>jobs:
  job01:
    name: Job 01
    runs-on: ubuntu-latest
    outputs:
      run_rest_jobs: ${{ steps.set_output.outputs.run_jobs }}
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          fetch-depth: 2

      - name: git diff and set output
        id: set_output
        run: |
          codeChanges="only within certain files"
          if [[$codeChanges == "only within certain files"]]; then
            echo "::set-output name=run_jobs::true"
          else
            echo "::set-output name=run_jobs::false"
          fi
  job02:
    name: Job 02
    needs: [job01]
    if: needs.job01.outputs.run_rest_jobs == 'true'
    runs-on: ubuntu-latest
    steps:
      - name: the rest jobs
        run: echo "run the rest jobs"
</code></pre>
<p>Description:</p>
<p><strong>1)</strong> On <strong>job01</strong> , in the " <strong>git diff and set output</strong>" step suppose the variable " <strong>codeChanges</strong>" is used to return the result of <strong>git diff</strong>. If the result confirms that the code changes only within certain files, set the step’s output parameter " <strong>run_jobs</strong>" to be " <strong>true</strong>", otherwise set " <strong>run_jobs</strong>" to be " <strong>false</strong>". The job’s output parameter " <strong>run_rest_jobs</strong>" will use the value of " <strong>run_jobs</strong>".</p>
<p><strong>2)</strong> On <strong>job02</strong> , use “<strong>needs: [job01]</strong>” to set only  <strong>job01</strong>  must complete successfully before <strong>job02</strong>  begins to run, and use the  <strong>if conditional</strong> ( <strong>if: needs.job01.outputs.run_rest_jobs == ‘true’</strong> ) to check whether skip  <strong>job02</strong>  or not.</p>
<p><strong>3)</strong> Similarly on other subsequent jobs.</p>
<p>Related docs:</p>
<p><a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjobs_idoutputs" rel="nofollow noopener"><strong>jobs.&lt;jobs_id&gt;.outputs</strong></a></p>
<p><a href="https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-output-parameter" rel="nofollow noopener"><strong>Setting an output parameter</strong></a></p>
<p><a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds" rel="nofollow noopener"><strong>jobs.&lt;job_id&gt;.needs</strong></a></p>
<p><a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idif" rel="nofollow noopener"><strong>jobs.&lt;job_id&gt;.if</strong></a></p>
<p><a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#steps-context" rel="nofollow noopener"><strong>steps context</strong></a></p>
<p><a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#needs-context" rel="nofollow noopener"><strong>needs context</strong></a></p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/blackswan2019">@blackswan2019</a> ,</p>
<p>How are things going? Is my above suggestion helpful to you?</p>
<p>If you have any question about this ticket, feel free to tell me.</p>

**Comment by BlackSwan2019** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a>. I haven’t got around to trying it yet, but it looks like it would work. Will update once implemented! Thanks.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/blackswan2019">@blackswan2019</a> ,</p>
<p>Okey, any progress, feel free to tell me.</p>

**Comment by blane1988** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a> , I have implemented the first part (similar to the job1), and the <code>need</code> requirement for job2 (the dependent job), but the <code>if</code> isn’t working. I noticed that the output from your job1 example is “run_jobs” whereas in job2, the if condition uses “run_rest_jobs”. Is there a reason they are different?</p>
<p>Overall, does my if condition look good? Thanks!</p>
<pre><code>if: needs.native-code-diff-check.outputs.run_ios_job == 'true'
</code></pre>
<p>EDIT: Okay, I echoed out the needs context in job2, and it looks like the previous job (job1) never even sent the output value. I will investigate why this is!</p>
<pre><code>NEEDS CONTEXT {
  "native-code-diff-check": {
    "result": "success",
    "outputs": {}
  }
}
</code></pre>
<p>EDIT 2: Solved it. It was a mismatch between run_jobs and run_ios_job in two lines respectively. </p>
<pre><code>outputs:
      run_rest_jobs: ${{ steps.native-files-check.outputs. **run\_jobs** }}

echo "::set-output name= **run\_ios\_job** ::true"
</code></pre>
<p>I set both to run_ios_job.</p>
<p>Your solution did the trick. Thanks for the great response and sample code/docs!</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/blane1988">@blane1988</a> ,</p>
<p>On job1, <strong>run_rest_jobs</strong> is a <strong>job-level output</strong> , <strong>run_jobs</strong> is a <strong>step-level output</strong>.</p>
<p>What you should know is, when job2 <strong>needs</strong> job1, job2 only can access job-level output from job1, and can’t directly access step-level output from job1. Because in the <strong>needs context</strong> , it only lists the job-level outputs on job1, and no expression syntax allows job2 to directly access the step-level outputs from job1 (such as " <strong>needs.job01.steps.set_output.outputs.run_jobs</strong>" is not supported).</p>
<p>So, on job2, if you want to get the value of the step-level output ( <strong>run_jobs</strong> ) from job1, you need to set a job-level output ( <strong>run_rest_jobs</strong> ), and pass the value of the step-level output to the job-level output. If job2 needs to access multiple step-level outputs on job1, you need to set a job-level output for each step-level output.</p>
<p>In order to easily distinguish between the job-level output and the step-level output, I give them different names. Of course, giving them the same name also is OK.</p>

**Comment by janosh** [comment]
<p>Is there still no simpler way to abort workflows? Something like</p>
<pre><code class="lang-bash">if [ "$condition" -ne "$value" ]; then
    echo 'Aborting workflow...'
    exit 0 # need some way to kill whole workflow, not just step
fi
</code></pre>

**Comment by BlackSwan2019** [comment]
<p>We wanted to abort job as well (and in our case, not consider it a failure), but no, we never found a way. The shell command <code>exit &lt;number&gt;</code> only applied to steps, like you commented. It’d be great if GitHub introduced some shell alias command that, when called, just skips the whole rest of the workflow.</p>
<p>So, instead, we just have every step in the job have an if-condition, that, when met, skips the steps in sequence until whole workflow is effectively skipped. But, yeah… not pretty.</p>

**Comment by bparthas21** [comment]
This is why we are moving away from coding in GH workflows. They are a pain. Just write everything in your own code and use GH Actions workflows just as a skeleton to call your code.

**Comment by brunoalr** [comment]
+1.
This would be a highly appreciated quality-of-life feature for GHA.
